plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts

```diff
@@ -65,8 +71,31 @@ export class AuthorizedEntitiesCatalog implements EntitiesCatalog {
     return this.entitiesCatalog.entities(request);
   }
 
-  removeEntityByUid(uid: string): Promise<void> {
-    // TODO: Implement permissioning
+  async removeEntityByUid(
+    uid: string,
+    authorizationToken?: string,
+  ): Promise<void> {
+    const { entities } = await this.entitiesCatalog.entities({
+      filter: basicEntityFilter({ 'metadata.uid': uid }),
+    });
+    if (entities.length === 0) {
+      throw new NotAllowedError();
+    }
+
+    const authorizeResponse = (
+      await this.permissionApi.authorize(
+        [
+          {
+            permission: catalogEntityDeletePermission,
+            resourceRef: stringifyEntityRef(entities[0]),
+          },
+        ],
+        { token: authorizationToken },
+      )
+    )[0];
+    if (authorizeResponse.result !== AuthorizeResult.ALLOW) {
+      throw new NotAllowedError();
```

Comment:
Does this error actually trigger a 403 to be sent back? I can't see that it does.

Reply:
It should be transformed in the [`errorHandler`](https://github.com/backstage/backstage/blob/532f8580f0800fae86fed9f5f4fa2182b418c214/packages/backend-common/src/middleware/errorHandler.ts#L129-L130) middleware. See usage in [`NextRouter`](https://github.com/backstage/backstage/blob/532f8580f0800fae86fed9f5f4fa2182b418c214/plugins/catalog-backend/src/service/NextRouter.ts#L203)

Reply:
FWIW the entity read endpoints will return a 404 if the requesting user doesn't have access

Reply:
Thanks for pointing that out! One thing @Rugvip mentioned is that if users have read access to a given entity (thus they know that it exists) and they try to delete it but don't have delete access, then it would be weird to tell that user the entity doesn't exist, when they know for a fact it does. But I do agree, this could also be used by a user who does not know the existence of an entity to gain information. I think here and in other endpoint PRs I've been generally leaning towards using `NotAllowedError` where it makes sense, but maybe this is a topic worth discussing offline if we feel otherwise.

Reply:
See the discussion [here](https://github.com/backstage/backstage/pull/8783#discussion_r779705609). TL;DR: The important thing is that we return the same result for both missing and restricted resources, and it's up to the permission handler to make sure that happens.

I think using `NotAllowedError` is the nicer way to go as that allows you to opt-in to having `NotFoundError`s if you want them, by returning `ALLOW` for missing resources.

Reply:
Won't that be hard to control from the permission backend for the conditional case? For example, if we make the authorize call with no `resourceRef` (discussed in https://github.com/backstage/backstage/pull/8805/files#r780390447) and it returns conditions, then we can't differentiate between missing and unauthorized entities.

@Joonpark13 that's a good point about users that can read but not delete an entity. A couple options that come to mind:
- Return a 403 from the delete endpoint for both cases (ie you aren't authorized to delete entities that don't exist).
- Return a 404 from the delete endpoint and have the frontend translate it to some "unauthorized" message.

Reply:
> then we can't differentiate between missing and unauthorized entities.

I think what @Rugvip is saying is that we shouldn't try to differentiate at all and just return `NotAllowedError` for that case.

> Return a 403 from the delete endpoint for both cases

It seems like this is the solution we are leaning towards.

----------
Overall, I think it's a bit strange to say that something is not allowed when it really doesn't exist and vice versa, but I do agree that the combined benefit of delegating to the policy + avoiding leaking information outweighs this clarity. In summary, no further change is needed.